src/loc.rs

```diff
@@ -0,0 +1,466 @@
+    where Endian: Endianity
+{
+    debug_loc_section: EndianBuf<'input, Endian>,
+}
+
+impl<'input, Endian> DebugLoc<'input, Endian>
+    where Endian: Endianity
+{
+    /// Construct a new `DebugLoc` instance from the data in the `.debug_loc`
+    /// section.
+    ///
+    /// It is the caller's responsibility to read the `.debug_loc` section and
+    /// present it as a `&[u8]` slice. That means using some ELF loader on
+    /// Linux, a Mach-O loader on OSX, etc.
+    ///
+    /// ```
+    /// use gimli::{DebugLoc, LittleEndian};
+    ///
+    /// # let buf = [0x00, 0x01, 0x02, 0x03];
+    /// # let read_debug_loc_section_somehow = || &buf;
+    /// let debug_loc = DebugLoc::<LittleEndian>::new(read_debug_loc_section_somehow());
+    /// ```
+    pub fn new(debug_loc_section: &'input [u8]) -> DebugLoc<'input, Endian> {
+        DebugLoc { debug_loc_section: EndianBuf(debug_loc_section, PhantomData) }
+    }
+
+    /// Iterate over the `LocationListEntry`s starting at the given offset.
+    ///
+    /// The `address_size` must be match the compilation unit for this location list.
+    /// The `base_address` should be obtained from the `DW_AT_low_pc` attribute in the
+    /// `DW_TAG_compile_unit` entry for the compilation unit that contains this location
+    /// list.
+    ///
+    /// Can be [used with
+    /// `FallibleIterator`](./index.html#using-with-fallibleiterator).
+    pub fn locations(&self,
+                     offset: DebugLocOffset,
+                     address_size: u8,
+                     base_address: u64)
+                     -> Result<LocationListIter<Endian>> {
+        let offset = offset.0 as usize;
+        if self.debug_loc_section.len() < offset {
+            return Err(Error::UnexpectedEof);
+        }
+
+        let input = self.debug_loc_section.range_from(offset..);
+        Ok(LocationListIter::new(input, address_size, base_address))
+    }
+
+    /// Iterate over the raw `LocationListEntry`s starting at the given offset.
+    ///
+    /// The `address_size` must be match the compilation unit for this location list.
+    ///
+    /// This iterator does not perform any processing of the location entries,
+    /// such as handling base addresses.
+    ///
+    /// Can be [used with
+    /// `FallibleIterator`](./index.html#using-with-fallibleiterator).
+    pub fn raw_locations(&self,
+                         offset: DebugLocOffset,
+                         address_size: u8)
+                         -> Result<RawLocationListIter<Endian>> {
+        let offset = offset.0 as usize;
+        if self.debug_loc_section.len() < offset {
+            return Err(Error::UnexpectedEof);
+        }
+
+        let input = self.debug_loc_section.range_from(offset..);
+        Ok(RawLocationListIter::new(input, address_size))
+    }
+}
+
+/// A raw iterator over a location list.
+///
+/// This iterator does not perform any processing of the location entries,
+/// such as handling base addresses.
+#[derive(Debug)]
```

Comment:
Yeah I think we should. There might be some pragmas we could use to help the compiler enforce this